verification/io/io-spec.gobra

```diff
@@ -92,79 +92,92 @@ pure func dp3s_iospec_bio3s_enter_guard(s IO_dp3s_state_local, t Place, v IO_val
-	return let currseg := v.IO_Internal_val1_1.CurrSeg in
-	v.IO_Internal_val1_1.LeftSeg == none[IO_seg2] ? false :
-	let nextseg := get(v.IO_Internal_val1_1.LeftSeg) in
-	(nextseg.ConsDir &&
-	len(nextseg.Future) > 0 &&
-	len(currseg.Future) > 0 &&
-	let hf1, hf2 := currseg.Future[0], nextseg.Future[0] in
-	let traversedseg := dp3s_iospec_bio3s_create_guard_traversedseg(currseg) in
-	(xover_up2down2_link_type(asid(), hf1, hf2) &&
-		dp3s_xover_common(
-			s,
-			v.IO_Internal_val1_1,
-			currseg,
-			nextseg,
-			traversedseg,
-			IO_pkt2(IO_Packet2{nextseg, v.IO_Internal_val1_1.MidSeg, v.IO_Internal_val1_1.RightSeg, some(traversedseg)}),
-			hf1,
-			hf2,
-			v.IO_Internal_val1_2,
-			v.IO_Internal_val1_3,
-			v.IO_Internal_val1_4,)))
+pure func (dp DataPlaneSpec) dp3s_iospec_bio3s_xover_up2down_guard(s IO_dp3s_state_local, t Place, v IO_val) bool {
+	return let currseg := v.IO_Internal_val1_1.CurrSeg in 
+		match v.IO_Internal_val1_1.LeftSeg{
+			case none[IO_seg2]: false
+			default: 
+				let nextseg := get(v.IO_Internal_val1_1.LeftSeg) in
+				(nextseg.ConsDir &&
+				len(nextseg.Future) > 0 &&
+				len(currseg.Future) > 0 &&
+				let hf1, hf2 := currseg.Future[0], nextseg.Future[0] in
+				let traversedseg := dp.dp3s_iospec_bio3s_create_guard_traversedseg(currseg) in
+				(dp.xover_up2down2_link_type(dp.asid(), hf1, hf2) &&
+					dp.dp3s_xover_common(
+						s,
+						v.IO_Internal_val1_1,
+						currseg,
+						nextseg,
+						traversedseg,
+						IO_pkt2(IO_Packet2{nextseg, v.IO_Internal_val1_1.MidSeg, v.IO_Internal_val1_1.RightSeg, some(traversedseg)}),
+						hf1,
+						hf2,
+						v.IO_Internal_val1_2,
+						v.IO_Internal_val1_3,
+						v.IO_Internal_val1_4,)))
+		}
 }
 
-pred dp3s_iospec_bio3s_xover_up2down(s IO_dp3s_state_local, t Place) {
-	forall v IO_val :: { dp3s_iospec_bio3s_xover_up2down_guard(s, t, v) }{ CBio_IN_bio3s_xover_up2down(t, v) } { dp3s_iospec_ordered(dp3s_add_obuf(s, v.IO_Internal_val1_4, v.IO_Internal_val1_3), dp3s_iospec_bio3s_xover_up2down_T(s, t, v)) } (v.isIO_Internal_val1 ==>
-		(dp3s_iospec_bio3s_xover_up2down_guard(s, t, v) ==>
-			(CBio_IN_bio3s_xover_up2down(t, v) &&
-				dp3s_iospec_ordered(
-					dp3s_add_obuf(s, v.IO_Internal_val1_4, v.IO_Internal_val1_3),
-					dp3s_iospec_bio3s_xover_up2down_T(s, t, v)))))
+pred (dp DataPlaneSpec) dp3s_iospec_bio3s_xover_up2down(s IO_dp3s_state_local, t Place) {
+	forall v IO_val :: { dp.dp3s_iospec_bio3s_xover_up2down_guard(s, t, v) }{ CBio_IN_bio3s_xover_up2down(t, v) } { dp.dp3s_iospec_ordered(dp.dp3s_add_obuf(s, v.IO_Internal_val1_4, v.IO_Internal_val1_3), dp3s_iospec_bio3s_xover_up2down_T(t, v)) } (
+		match v {
+			case IO_Internal_val1{_, _, ?newpkt, ?nextif}: 
+				(dp.dp3s_iospec_bio3s_xover_up2down_guard(s, t, v) ==>
+				(CBio_IN_bio3s_xover_up2down(t, v) &&
+					dp.dp3s_iospec_ordered(
+						dp.dp3s_add_obuf(s, nextif, newpkt),
+						dp3s_iospec_bio3s_xover_up2down_T(t, v))))
+			default: true
+		})
```

Comment:
Do we need the parentheses surrounding the match?

Reply:
Gobra forces you to write right after the triggers. So either we drop the parentheses and move `match v {` one line up or leave the parentheses. I find the latter one more readable.